### PR TITLE
hmac-sha512 support added to httpsignature

### PR DESCRIPTION
**Pull request by evq**
**NOTE** PR description + base copied from: https://github.com/brave-intl/bat-go/pull/890

### Summary

Added fully HS2019 (HMAC-SHA512) support to the `httpsigning` package
I have added tests and verified that the tests work
I have tried to add comments on the code that I changed

Meant to be reviewed by @evq based on private discussions in Slack

### Type of change ( select one )

- [x] Product feature
- [ ] Bug fix
- [ ] Performance improvement
- [ ] Refactor
- [ ] Other

### Tested Environments

- [x] Development
- [ ] Staging
- [ ] Production

### Before submitting this PR:

- [x] Does your code build cleanly without any errors or warnings?
- [ ] Have you used auto closing keywords?
- [x] Have you added tests for new functionality?
- [ ] Have validated query efficiency for new database queries?
- [ ] Have documented new functionality in README or in comments?
- [ ] Have you squashed all intermediate commits?
- [x] Is there a clear title that explains what the PR does?
- [x] Have you used intuitive function, variable and other naming?
- [ ] Have you requested security / privacy review if needed
- [x] Have you performed a self review of this PR?

### CODE REVIEW

**Comment by tzeejay** [comment]
@evq This looks great to me. I think this is going to serve us very well.

Only thing that I would personally add would be a `VerifyRequest()` method just to keep things obvious since a `SignRequest()` method exists. Even if it calls directly into `Verify()` like `SignRequest()` does. That way there are two pairs of methods that work in tandem.

File: utils/httpsignature/hmac.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+	"crypto/sha512"
+	"crypto/subtle"
+	"io"
+)
+
+// HMACKey is a symmetric key that can be used for HMAC-SHA512 request signing and verification
+type HMACKey string
+
+// Sign the message using the hmac key
+func (key HMACKey) Sign(rand io.Reader, message []byte, opts crypto.SignerOpts) (signature []byte, err error) {
```
Comment (orspetol) [inline utils/httpsignature/hmac.go:15]: Why is `rand` passed into this method, as well as the interface? There is no usage of it that I can see. Unless the intended interface was meant to abstract any type of signature that may require rand? Regardless my general recommendation is to never allow higher level functions to pass rand to the cryptographic algo due to api abuse (far too often callers passing rand are not sending enough - or any entropy).

File: utils/httpsignature/hmac.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+// Verify the signature sig for message using the hmac key
+func (key HMACKey) Verify(message, sig []byte, opts crypto.SignerOpts) (bool, error) {
+	hashSum, err := key.Sign(nil, message, nil)
+	if err != nil {
+		return false, err
+	}
+	// Return bool by checking whether or not the calculated hash is equal to
+	// sig pulled out of the header. Check if returned int is equal to 1 to return a bool
+	return subtle.ConstantTimeCompare(hashSum, sig) == 1, nil
```
Comment (orspetol) [inline utils/httpsignature/hmac.go:34]: Thank you

File: utils/httpsignature/httpsignature.go [hunk 3]
```diff
@@ -20,15 +21,34 @@ import (
...
+// signature is an internal represention of an http signature and it's parameters
+type signature struct {
 	SignatureParams
-	Sig string // FIXME remove since we should always use http.Request header?
+	Sig string
+}
+
+// Signator is an interface for cryptographic signature creation
+type Signator interface {
+	Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error)
```
Comment (orspetol) [inline utils/httpsignature/httpsignature.go:38]: re my comment above: I suppose this `Signator` interface may need rand exposed due to signator type, but I still consider it risky.

File: wallet/controllers.go [hunk 4]
```diff
@@ -43,8 +43,7 @@ func (service *Service) LookupPublicKey(ctx context.Context, keyID string) (*htt
 
 func validateHTTPSignature(ctx context.Context, r *http.Request, signature string) (string, error) {
 	// validate that the signature in the header is valid based on public key provided
-	var s httpsignature.Signature
-	err := s.UnmarshalText([]byte(signature))
+	s, err := httpsignature.SignatureParamsFromRequest(r)
```
Comment (orspetol) [inline wallet/controllers.go:46]: Hopefully the right layer: should we validate `SignatureParamsFromRequest` length? Im not sure off the top of my head if there is a max acceptable length? Or if that is even a concern (I cant image much security risk)

File: utils/httpsignature/httpsignature.go (hunk 3 above)
Comment (evq) [reply utils/httpsignature/httpsignature.go:38]: I probably should have an inline comment noting that this is basically a subset of the https://pkg.go.dev/crypto#Signer interface. I made the choice to use `Sign` from the standard library `crypto.Signer` rather a custom function signature a while back - by preserving it here I was trying to make sure we wouldn't  have to make changes to existing call sites

**Review by orspetol — DISMISSED** [review]

**Review by orspetol — APPROVED** [review]

File: utils/httpsignature/httpsignature.go (hunk 3 above)
Comment (evq) [reply utils/httpsignature/httpsignature.go:40]: Added a comment

File: wallet/controllers.go (hunk 4 above)
Comment (evq) [reply wallet/controllers.go]: Added an explicit check for size = 0

File: middleware/http_signed_test.go [hunk 5]
```diff
@@ -56,7 +56,7 @@ func TestHTTPSignedOnly(t *testing.T) {
 	assert.NoError(t, err)
 	rr = httptest.NewRecorder()
 	handler.ServeHTTP(rr, req)
-	assert.Equal(t, http.StatusNotFound, rr.Code, "request with signature from wrong keyID should fail")
+	assert.Equal(t, http.StatusForbidden, rr.Code, "request with signature from wrong keyID should fail")
```
Comment (husobee) [inline middleware/http_signed_test.go:59]: :+1:  StatusForbidden is good when using an invalid key

File: middleware/http_signed_test.go [hunk 6]
```diff
@@ -41,9 +41,9 @@ func TestHTTPSignedOnly(t *testing.T) {
 	assert.NoError(t, err)
 	rr := httptest.NewRecorder()
 	handler.ServeHTTP(rr, req)
-	assert.Equal(t, http.StatusBadRequest, rr.Code, "request without signature should fail")
+	assert.Equal(t, http.StatusForbidden, rr.Code, "request without signature should fail")
```
Comment (husobee) [inline middleware/http_signed_test.go:44]: feel like this should be StatusUnauthorized instead, as the request lacks valid authentication credentials for the target resource.

**Review by husobee — APPROVED** [review]
looks good, response code change in the event no credentials are sent might be good.

**Review by husobee — APPROVED** [review]

**Pull request merged**